src/lib/pickles/wrap.ml

```diff
@@ -283,518 +283,391 @@ let deferred_values (type n) ~(sgs : (Backend.Tick.Curve.Affine.t, n) Vector.t)
+            Step_verifier.finalize_other_proof
+              (module Nat.N0)
+              ~feature_flags
+              ~step_domains:
+                (`Known
+                  [ { h = Pow_2_roots_of_unity vk.domain.log_size_of_group } ]
+                  )
+              ~sponge ~prev_challenges:[] deferred_values evals
+          in
+
+          (* Actually do the above as the prover *)
+          Impls.Step.(As_prover.(fun () -> read Boolean.typ res)) )
+      |> Or_error.ok_exn
+
+    (* Common srs value for all tests *)
     let srs =
       Kimchi_bindings.Protocol.SRS.Fp.create (1 lsl Common.Max_degree.step_log2)
-    in
-    let index, public_input, proof =
-      Kimchi_bindings.Protocol.Proof.Fp.example_with_ffadd srs
-    in
-
-    (* Obtain verifier key from prover index and convert backend proof to snarky proof *)
-    let vk = Kimchi_bindings.Protocol.VerifierIndex.Fp.create index in
-    let proof = Backend.Tick.Proof.of_backend proof in
 
-    (* Specify feature flags that were used for backend proof *)
-    let actual_feature_flags =
-      let open Plonk_types.Opt.Flag in
-      { Plonk_types.Features.chacha = false
-      ; range_check = true
-      ; foreign_field_add = true
-      ; foreign_field_mul = false
-      ; xor = false
-      ; rot = false
-      ; lookup = true
-      ; runtime_tables = false
-      }
-    in
-
-    (* Run the recursive proof test with supplied feature flags *)
-    run_custom_gate_tests actual_feature_flags vk proof
-  with _e ->
-    Printexc.print_backtrace stdout ;
-    Out_channel.flush stdout ;
-    exit 2
+    type example =
+         Kimchi_bindings.Protocol.SRS.Fp.t
+      -> Kimchi_bindings.Protocol.Index.Fp.t
+         * Pasta_bindings.Fp.t list
+         * ( Pasta_bindings.Fq.t Kimchi_types.or_infinity
+           , Pasta_bindings.Fp.t )
+           Kimchi_types.prover_proof
+
+    module type SETUP = sig
+      val example : example
+
+      (* Feature flags tused for backend proof *)
+      val actual_feature_flags : bool Plonk_types.Features.t
+    end
+
+    (* [Make] is the test functor.
+
+       Given a test setup, compute different test configurations and define 3
+       test for said configurations. *)
+    module Make (S : SETUP) = struct
+      (* Generate foreign field multiplication test backend proof using Kimchi,
+         obtaining the proof and corresponding prover index.
+
+         Note: we only want to pay the cost of generating this proof once and
+         then reuse it many times for the different recursive proof tests. *)
+      let index, public_input, proof = S.example srs
+
+      (* Obtain verifier key from prover index and convert backend proof to
+         snarky proof *)
+      let vk = Kimchi_bindings.Protocol.VerifierIndex.Fp.create index
+
+      let proof = Backend.Tick.Proof.of_backend proof
+
+      let test_feature_flags_configs =
+        generate_test_feature_flag_configs S.actual_feature_flags
+
+      let runtest feature_flags =
```

Comment:
why dont we normally use parenthesis in ocaml to pass the parameters as in other languages?

Reply:
🤯 Okay this is very useful learning, thanks for sharing!